public/js/src/module/form-cache.js

```diff
@@ -25,21 +44,80 @@ function init( survey ) {
         .then( _setResetListener );
 }
 
+/**
+ * @param {Survey} survey
+ * @return Survey
+ */
 function get( survey ) {
     return store.survey.get( survey.enketoId );
 }
 
+/**
+ * @param {Survey} survey
+ * @return {Promise<Survey>}
+ */
 function set( survey ) {
     return connection.getFormParts( survey )
         .then( _swapMediaSrc )
         .then( _addBinaryDefaultsAndUpdateModel )
         .then( store.survey.set );
 }
 
+/**
+ *
+ * @param { string } enketoId
+ * @return { Promise<EnketoRecord | undefined> }
+ */
+function getLastSavedRecord( enketoId ) {
+    return store.survey.get( enketoId )
+        .then( survey => {
+            if ( survey != null ) {
+                return survey.lastSavedRecord;
+            }
+        } );
+}
+
+/**
+ *
+ * @param { string } enketoId
+ * @param { EnketoRecord } lastSavedRecord
+ * @return { Promise<Survey> }
+ */
+function setLastSavedRecord( enketoId, lastSavedRecord ) {
+    return store.survey.get( enketoId )
+        .then( survey => {
+            if ( encryptor.isEncryptionEnabled( survey ) ) {
+                return Promise.resolve( survey );
+            }
+
+            if ( Array.isArray( survey.externalData ) ) {
+                const { xml } = lastSavedRecord;
+
+                survey.externalData.forEach( item => {
```

Comment:
It actually just didn't occur to me to use `find` here! You're right about the performance impact, but I think using `find` may actually make this easier to understand. And it would also be a better place to ensure `lastSavedRecord` isn't stored when the form doesn't reference it.

Reply:
As far as other usage of `forEach` I think they're mostly in tests, where the intent was to test deep equality over a subset of properties. Although I see lodash is already in use in the project, so I may switch those to `_.pick` calls if you think that would make those cases easier to understand?

I tend to avoid lodash except when it's already present and improves code readability, but this may be one of those cases. Not because I'm one of those "you don't need a utility library" purists, but because lodash has quite a few footguns (e.g. confusing argument order, surprising implicit behavior when some arguments are omitted, unexpected mutation).

Reply:
> I tend to avoid lodash 

Yes, I do too. The existing use of lodash is probably no longer necessary either.